### PR TITLE
[REG-1844] Add ActionComplete keyFrameCriteria

### PR DESCRIPTION
**Pull request by abeizer**
Adds new bot segment criteria `ActionComplete` which will wait until a BotAction is complete. For a monobehaviour, this means calling `Destroy` on itself so that the action knows when to flag itself as "done". 

This allows us to move ahead with the blue flower test in the short term, but we may need to consider allowing the behaviour to report a pass/failure state to interrupt the sequence if something goes wrong.

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BehaviourCompleteKeyFrameCriteriaData.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+using System;
+using System.Text;
+using RegressionGames.StateRecorder.JsonConverters;
+
+namespace RegressionGames.StateRecorder.BotSegments.Models
+{
+    [Serializable]
+    public class BehaviourCompleteKeyFrameCriteriaData : IKeyFrameCriteriaData
+    {
+        public int apiVersion = SdkApiVersion.VERSION_1;
```
Comment (RG-nAmKcAz) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BehaviourCompleteKeyFrameCriteriaData.cs:10]: This needs a new SdkApiVersion defined with a line comment in `SdkApiVersion` describing when this support was added

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/OrKeyFrameCriteriaEvaluator.cs [hunk 2]
```diff
@@ -5,13 +5,14 @@ namespace RegressionGames.StateRecorder.BotSegments
 {
     public static class OrKeyFrameCriteriaEvaluator
     {
-        public static bool Matched(bool firstSegment, int segmentNumber, KeyFrameCriteria criteria)
+        public static bool Matched(bool firstSegment, BotSegment botSegment, KeyFrameCriteria criteria)
```
Comment (RG-nAmKcAz) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/OrKeyFrameCriteriaEvaluator.cs:8]: I had intentionally not passed the full BotSegment reference in this methods to limit data access and avoid the potential ambiguity in introduces.  By passing this, implements can now access the entire raw key frame criteria from the segment instead of being limited to the one criteria being passed.  Let's undo this part of the changes please.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs [hunk 3]
```diff
@@ -80,13 +80,15 @@ public void PersistPriorFrameStatus()
          * <summary>Publicly callable.. caches the statuses of the last passed key frame for computing delta counts from</summary>
          */
 
-        public bool Matched(bool firstSegment, int segmentNumber, List<KeyFrameCriteria> criteriaList)
+        public bool Matched(bool firstSegment, BotSegment botSegment)
```
Comment (RG-nAmKcAz) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs:83]: This method accepting BotSegment seems fine/good, but MatchedHelper and other methods and evaluator class methods should only get the segmentNumber like before these changes as noted in my other comment.

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]

**Comment by abeizer** [comment]
@RG-nAmKcAz Updated - back to you

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs [hunk 4]
```diff
@@ -184,6 +182,13 @@ internal bool MatchedHelper(bool firstSegment, int segmentNumber, BooleanCriteri
                      case KeyFrameCriteriaType.CVImage:
                         cvImagesToMatch.Add(entry);
                         break;
+                    case KeyFrameCriteriaType.BehaviourComplete:
```
Comment (RG-nAmKcAz) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs:185]: Now that I see this all put together.. what if we made this `ActionComplete` , would that work / be more powerful so that we can just wait for any segment action to complete ?

**Review by RG-nAmKcAz — COMMENTED** [review]
One question before approving... which may set us up better for future use cases.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs (hunk 4 above)
Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs:185]: I'll +1 this idea of making it more generic so we can use it elsewhere!

**Review by vontell — APPROVED** [review]
Nice! Nothing in addition to what Zack already has commented here. Note that we should make sure to document this, but that might need to wait until we add a new section to the docs for all of this work anyways.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs [hunk 5]
```diff
@@ -182,7 +182,7 @@ internal bool MatchedHelper(bool firstSegment, int segmentNumber, BotActionType
                      case KeyFrameCriteriaType.CVImage:
                         cvImagesToMatch.Add(entry);
                         break;
-                    case KeyFrameCriteriaType.BehaviourComplete:
+                    case KeyFrameCriteriaType.ActionComplete:
                         if (botActionType == BotActionType.Behaviour && !botActionCompleted)
```
Comment (RG-nAmKcAz) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/KeyFrameEvaluator.cs:186]: We don't need the action type check here anymore

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]
1 small fix remaining

And.. I gave you a merge conflict in SdkApiVersion .... sorry

**Comment by abeizer** [comment]
@RG-nAmKcAz should all be fixed now. Back to you for hopefully the final time :-P

**Review by RG-nAmKcAz — APPROVED** [review]
